src/workerd/api/crypto/spkac.c++

```diff
@@ -1,102 +1,63 @@
-  auto pkey = NETSCAPE_SPKI_get_pubkey(spki.get());
-  if (!pkey) return kj::none;
-  return kj::disposeWith<EVP_PKEY_free>(pkey);
-}
-
-kj::Maybe<kj::Own<BIO>> tryNewBio() {
-  auto bioptr = BIO_new(BIO_s_mem());
-  if (!bioptr) return kj::none;
-  return kj::disposeWith<BIO_free_all>(bioptr);
-}
-}  // namespace
 
 bool verifySpkac(kj::ArrayPtr<const kj::byte> input) {
   // So, this is fun. SPKAC uses MD5 as the digest algorithm. This is a problem because
   // using MD5 for signature verification is not allowed in FIPS mode, which means that
   // although we have a working implementation here, the result of this call is always
-  // going to false even if the input signature is correct. So this is a bit of a dead
-  // end that isn't going to be super useful. Fortunately but the exportPublicKey and
+  // going to be false even if the input signature is correct. So this is a bit of a dead
+  // end that isn't going to be super useful. Fortunately tho the exportPublicKey and
   // exportChallenge functions both work correctly and are useful. Unfortunately, this
   // likely means users would need to implement their own verification, which sucks.
   //
   // Alternatively we could choose to implement our own version of the validation that
   // bypasses BoringSSL's FIPS configuration. For now tho, this does end up matching
   // Node.js' behavior when FIPS is enabled so I guess that's something.
-  ClearErrorOnReturn clearErrorOnReturn;
   if (IoContext::hasCurrent()) {
     IoContext::current().logWarningOnce(
         "The verifySpkac function is currently of limited value in workers because "
         "the SPKAC signature verification uses MD5, which is not supported in FIPS mode. "
         "All workers run in FIPS mode. Accordingly, this method will currently always "
         "return false even if the SPKAC signature is valid. This is a known limitation.");
   }
-  KJ_IF_SOME(spki, tryGetSpki(input)) {
-    KJ_IF_SOME(key, tryOwnPkey(spki)) {
-      return NETSCAPE_SPKI_verify(spki.get(), key.get()) > 0;
-    }
-  }
-  return false;
+
+  ncrypto::Buffer<const char> buf{
+    .data = reinterpret_cast<const char*>(input.begin()),
+    .len = input.size(),
+  };
+  return ncrypto::VerifySpkac(buf);
 }
 
 kj::Maybe<jsg::BufferSource> exportPublicKey(jsg::Lock& js, kj::ArrayPtr<const kj::byte> input) {
-  ClearErrorOnReturn clearErrorOnReturn;
-  KJ_IF_SOME(spki, tryGetSpki(input)) {
-    KJ_IF_SOME(bio, tryNewBio()) {
-      KJ_IF_SOME(key, tryOwnPkey(spki)) {
-        if (PEM_write_bio_PUBKEY(bio.get(), key.get()) > 0) {
-          return toArray(js, bio.get());
-        }
-      }
-    }
+  ncrypto::Buffer<const char> buf{
+    .data = reinterpret_cast<const char*>(input.begin()),
+    .len = input.size(),
+  };
+  if (auto bio = ncrypto::ExportPublicKey(buf)) {
+    BUF_MEM* bptr = bio;
+    auto buf = jsg::BackingStore::alloc(js, bptr->length);
+    auto aptr = kj::arrayPtr(bptr->data, bptr->length);
```

Comment:
Do you need to check for length/emptiness for safety?

Reply:
For these, we shouldn't. The copy operation that follows should be doing the right thing.